src/softmax.jl

```diff
@@ -61,23 +75,26 @@ It is semantically equivalent to the following:
 See also [`softmax`](@ref).
 """
 logsoftmax(x; dims = 1) = logsoftmax!(similar(x, (float ∘ eltype)(x)), x; dims = dims)
+
 logsoftmax!(x; dims = 1) = logsoftmax!(x, x; dims = dims)
+
 function logsoftmax!(out::O, x::T; dims = 1) where {O<:AbstractArray,T<:AbstractArray}
     out .= x .- maximum(x; dims = dims)
-    # out .-= log.(sum(exp.(out); dims = dims))  # WARN: this will decrease performance.
-    log_ = log.(sum(exp.(out); dims = dims))
-    out .-= log_
+    red = sum(exp, out; dims = dims)
+    red .= log.(red)
+    out .-= red
 end
 
-∇logsoftmax(Δ, x; dims = 1) = ∇logsoftmax!(similar(Δ), Δ, x; dims = dims)
-∇logsoftmax!(Δ, x; dims = 1) = ∇logsoftmax!(Δ, Δ, x; dims = dims)
-function ∇logsoftmax!(
-    out::O,
-    Δ::O,
-    x::T;
-    dims = 1,
-) where {O<:AbstractArray,T<:AbstractArray}
-    out .= Δ .- sum(Δ, dims = dims) .* softmax!(out, x; dims = dims)
+
+∇ logsoftmax(Δ:: AbstractArray{T}, x:: AbstractArray, y:: AbstractArray{S}; dims = 1) where {T,S} = 
+    ∇ logsoftmax!(similar(y, promote_type(T, S)), Δ, x, y; dims = dims)
```

Comment:
I think the spaces are my copy-paste error, sorry!
```suggestion
∇logsoftmax(Δ:: AbstractArray{T}, x:: AbstractArray, y:: AbstractArray{S}; dims = 1) where {T,S} = 
    ∇logsoftmax!(similar(y, promote_type(T, S)), Δ, x, y; dims = dims)
```